### PR TITLE
[LW-10504]add timeout transaction inspectors

### PR DESCRIPTION
**Pull request by VanessaPC**
# Context

The `transactionSummaryInspector` and `tokenTransferInspector` do not return a timeout error currently, when no inputs are resolved.  

# Proposed Solution
We add a `timeout` prop so the consumer can set up the time in MS in which they want the timeout to return if no inputs are found. 

# Important Changes Introduced
`transactionSummaryInspector` and `tokenTransferInspector` now require a `timeout` prop.

### CODE REVIEW

**Review by DominikGuzei — APPROVED** [review]

File: packages/core/src/util/tokenTransferInspector.ts [hunk 1]
```diff
@@ -148,11 +157,37 @@ const toTokenTransferValue = async (
...
+ * Resolve inputs with a timeout so that if the inputs fail to resolve in n Miliseconds
+ * we throw an error so the UI consumer can handle situation accordingly.
+ * In this case we create a race condition in which we race to send the resolved inputs before the
+ * timeout, and timeout only if the inputs aren't there.
+ */
+export const resolveWithTimeout = <T>(promise: Promise<T>, timeout: Milliseconds): Promise<T> =>
+  Promise.race([
+    promise,
+    new Promise<T>((_, reject) =>
+      setTimeout(() => reject(new Error('Timeout: failed to resolve inputs in time')), timeout)
```
Comment (mkazlauskas) [inline packages/core/src/util/tokenTransferInspector.ts:170]: A more specifically-typed error would be nice
```suggestion
      setTimeout(() => reject(new TimeoutError('Timeout: failed to resolve inputs in time')), timeout)
```

File: packages/core/src/util/tokenTransferInspector.ts [hunk 2]
```diff
@@ -148,11 +157,37 @@ const toTokenTransferValue = async (
...
+  ({ inputResolver, fromAddressAssetProvider, toAddressAssetProvider, timeout, logger }) =>
   async (tx) => {
-    const { resolvedInputs } = await resolveInputs(tx.body.inputs, inputResolver);
+    let resolvedInputs: ResolutionResult['resolvedInputs'];
+
+    try {
+      const inputResolution = await resolveWithTimeout(resolveInputs(tx.body.inputs, inputResolver), timeout);
+      resolvedInputs = inputResolution.resolvedInputs;
+    } catch (error) {
+      if (error === 'Timeout: failed to resolve inputs in time') {
```
Comment (mkazlauskas) [inline packages/core/src/util/tokenTransferInspector.ts:184]:
```suggestion
      if (error instanceof TimeoutError) {
```

File: packages/core/src/util/transactionSummaryInspector.ts [hunk 3]
```diff
@@ -147,8 +151,34 @@ const toAssetInfoWithAmount = async (
  */
 export const transactionSummaryInspector: TransactionSummaryInspector =
   (args: TransactionSummaryInspectorArgs) => async (tx) => {
-    const { inputResolver, addresses, rewardAccounts, protocolParameters, assetProvider, dRepKeyHash } = args;
-    const resolvedInputs = await resolveInputs(tx.body.inputs, inputResolver);
+    const {
+      inputResolver,
```
Comment (mkazlauskas) [inline packages/core/src/util/transactionSummaryInspector.ts:155]: We should memoize inputResolver in lace as both inspectors are calling it with same inputs

File: packages/core/test/util/tokenTransferInspector.test.ts [hunk 4]
```diff
@@ -221,6 +223,8 @@ describe('txInspector', () => {
 
   const assetProvider = createMockAssetProvider(assetInfos);
 
+  const timeout = 6000 as Milliseconds;
```
Comment (mkazlauskas) [inline packages/core/test/util/tokenTransferInspector.test.ts:226]: Missing timeout test for both inspectors

File: packages/core/src/util/transactionSummaryInspector.ts [hunk 5]
```diff
@@ -147,8 +151,34 @@ const toAssetInfoWithAmount = async (
...
+
+    let resolvedInputs: ResolutionResult;
+
+    try {
+      // eslint-disable-next-line sonarjs/prefer-immediate-return
+      resolvedInputs = await resolveWithTimeout(resolveInputs(tx.body.inputs, inputResolver), timeout);
+    } catch (error) {
+      if (error === 'Timeout: failed to resolve inputs in time') {
+        // Handle timeout error specifically
+        logger?.log('Error: Inputs resolution timed out');
```
Comment (mkazlauskas) [inline packages/core/src/util/transactionSummaryInspector.ts:173]: Is `logger` optional? It shouldn't be
```suggestion
        logger.log('Error: Inputs resolution timed out');
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/core/src/util/transactionSummaryInspector.ts (hunk 5 above)
Comment (przemyslaw-wlodek) [reply packages/core/src/util/transactionSummaryInspector.ts]: ✅

File: packages/core/test/util/tokenTransferInspector.test.ts (hunk 4 above)
Comment (przemyslaw-wlodek) [reply packages/core/test/util/tokenTransferInspector.test.ts:230]: ✅

File: packages/core/src/util/tokenTransferInspector.ts (hunk 1 above)
Comment (przemyslaw-wlodek) [reply packages/core/src/util/tokenTransferInspector.ts]: ✅

File: packages/core/test/util/tokenTransferInspector.test.ts [hunk 6]
```diff
@@ -1049,5 +1073,75 @@ describe('txInspector', () => {
         ])
       );
     });
+
+    it.only('has no resolved fromAddresses on timeout', async () => {
```
Comment (mkazlauskas) [inline packages/core/test/util/tokenTransferInspector.test.ts:1077]:
```suggestion
    it('has no resolved fromAddresses on timeout', async () => {
```

File: packages/core/test/util/transactionSummaryInspector.test.ts [hunk 7]
```diff
@@ -1017,4 +1043,81 @@ describe('Transaction Summary Inspector', () => {
       }
     });
   });
+
+  it('displays unresolved inputs on timeout', async () => {
+    const resolveDelay = 5;
```
Comment (mkazlauskas) [inline packages/core/test/util/transactionSummaryInspector.test.ts:1048]: No a required change, just a nitpick for slightly better test readability
```suggestion
    const resolveDelay = Milliseconds(5);
```

File: packages/core/src/util/tokenTransferInspector.ts [hunk 8]
```diff
@@ -150,9 +161,20 @@ const toTokenTransferValue = async (
 
 /** Inspect a transaction and return a map of addresses and their balances. */
 export const tokenTransferInspector: TokenTransferInspector =
```
Comment (mkazlauskas) [inline packages/core/src/util/tokenTransferInspector.ts:163]: I've just noticed that this PR only added timeout for input resolvers, but forgot to handle failures and timeouts of AssetProvider (e.g. https://github.com/input-output-hk/cardano-js-sdk/blob/b3f3280c9ddc0b9c0122c9c4be0eacc2201a7faf/packages/core/src/util/tokenTransferInspector.ts#L142 which can also fail/timeout) which is in scope of LW-10504

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/core/test/util/transactionSummaryInspector.test.ts (hunk 7 above)
Comment (przemyslaw-wlodek) [reply packages/core/test/util/transactionSummaryInspector.test.ts]: Fixed by 87baffa8fd705826f0073bec5d318204a87344e5

File: packages/core/test/util/tokenTransferInspector.test.ts (hunk 6 above)
Comment (przemyslaw-wlodek) [reply packages/core/test/util/tokenTransferInspector.test.ts]: Apologies, it should be fixed by 87baffa8fd705826f0073bec5d318204a87344e5 :)

File: packages/core/src/util/tokenTransferInspector.ts (hunk 8 above)
Comment (przemyslaw-wlodek) [reply packages/core/src/util/tokenTransferInspector.ts:175]: Should be fixed by 6b56dce807feb76cff36b051a10c36da0d584d4c

**Review by mkazlauskas — APPROVED** [review]
Nice work! :rocket:

**Review by mkazlauskas — APPROVED** [review]

**Pull request merged**